### PR TITLE
MultiJobCleanupJob

### PR DESCRIPTION
**Pull request by JackTemaki**
A job that can, depending on the mode, clean jobs or parts of jobs when a certain path gets available.

### CODE REVIEW

**Review by michelwi — COMMENTED** [review]
I decline this PR from a principle standpoint. a Job should be an atomic unit that only modifies its own folder and never the state of anything else in the graph (or other graphs, or other parts of the filesystem)

File: tools/graph.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        :param trigger: trigger Path, this should be the path that causes the MultiCleanUpJob to run
+            All the outputs/jobs to be deleted should be dependencies.
+        :param mode: what kind of deletion to perform
+        """
+        self.job_output_list = job_output_list
+        self.trigger = trigger
+        self.mode = mode
+        self.out = self.output_path(os.path.basename(trigger.path))
+
+        # check that the trigger does not share a common job with the job outputs to delete
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply tools/graph.py:41]: Hmm yes maybe I assert this explicitly for any mode that is not `output_file_only`.

File: tools/graph.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                        or self.mode is self.CleanupMode.work_and_output_folder
+                    ):
+                        shutil.rmtree(
+                            os.path.join(
+                                job_output.creator._sis_path(abspath=True), "output"
+                            )
+                        )
+            except Exception as e:
+                print("Deletion not possible with:")
+                print(e)
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply tools/graph.py:79]: Maybe I should add a flag for this, I had cases where I deleted the jobs by hand earlier than the job here if possible, so I guess this is a leftover from that. But I maybe this is not needed any more and I can always raise the Exception.

File: tools/graph.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+class MultiJobCleanupJob(Job):
+    """
+    This Job will remove all given job folders when the provided Path is available.
+
+    Note that the provided jobs should not have any directly registered output, as otherwise they
+    will re-run right away.
+    """
+
+    class CleanupMode(enum.Enum):
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply tools/graph.py:19]: Yes and yes.

**Comment by JackTemaki** [comment]
> I decline this PR from a principle standpoint. a Job should be an atomic unit that only modifies its own folder and never the state of anything else in the graph (or other graphs, or other parts of the filesystem)

I understand this concern, unfortunately me and my Hiwis have experiments that are impossible to run without this job. We can keep this locally of course. I do not see a simple solution to run the sis cleaner automatically instead.

**Comment by curufinwe** [comment]
What about adding a separate functionality to sisyphus that marks jobs as deletable or compressable once all dependent jobs are finished. That would be much cleaner in my view.

**Comment by JackTemaki** [comment]
Okay I will drop this PR, but most likely not find the time to do a Sisyphus side implementation of this soon.

**Comment by michelwi** [comment]
I think one problem would be that if a job is once finished all outputs are assumed to exist. So if we cleanup only some files in the job directory and then later add another output this will break.
Deleting the whole job folder if all outputs are finished would be save. In that case if new outputs are added the job is automatically rerun.

**Comment by michelwi** [comment]
Independent of whether we want this or not, here is also a way to achieve that functionality using sisyphus callbacks:

````
import os

import i6_core.tools as tools
import i6_core.text as text

from sisyphus import tk


def delete_file(path: tk.Path, dependency: tk.Path):
    assert os.path.exists(dependency.get_path())
    if os.path.exists(path.get_path()):
        os.unlink(path.get_path())


def py():

    returnn_root = tools.CloneGitRepositoryJob(
        url="https://github.com/rwth-i6/returnn.git",
    ).out_repository

    modified_readme = text.PipelineJob(
        text_files=returnn_root.join_right("README.rst"),
        pipeline=["sed 's/RWTH/AppTek/g'"],
    ).out

    tk.register_output("README.rst", modified_readme)
    tk.register_callback(
        delete_file, returnn_root.join_right("README.rst"), modified_readme
    )
````

**Pull request closed without merging**